File: strands-py/src/strands/storage/local_file_storage.py
"""Local filesystem storage implementation."""

from __future__ import annotations

import builtins
import os
import uuid
from pathlib import Path
from typing import TYPE_CHECKING

from ..types.exceptions import StorageError
from .storage import _NamespacedStorage, _normalize_key, _normalize_prefix

if TYPE_CHECKING:
    from ..sandbox.base import Sandbox

_TMP_MARKER = ".__strands_tmp"


class LocalFileStorage:
    """Persists each key as a file under a base directory.

    Key segments separated by '/' map to directory segments. Writes on the host
    filesystem are atomic (write to temp file, then rename).

    Example:
        ```python
        from strands.storage import LocalFileStorage

        storage = LocalFileStorage("./.strands/")
        await storage.write("session/abc/state.json", data)
        ```
    """

    def __init__(self, base_dir: str = "./.strands/", *, sandbox: Sandbox | None = None) -> None:
        """Initialize local file storage.

        Args:
            base_dir: Root directory under which all keys are stored.
            sandbox: Optional sandbox to route I/O through.
        """
        self._base_dir = base_dir
        self._sandbox = sandbox

    def for_sandbox(self, sandbox: Sandbox) -> LocalFileStorage:
        """Return a copy bound to the given sandbox.

        If already bound to the same sandbox, returns self.

        Args:
            sandbox: Sandbox to bind to.

        Returns:
            A LocalFileStorage instance bound to the sandbox.
        """
        if self._sandbox is sandbox:
            return self
        return LocalFileStorage(self._base_dir, sandbox=sandbox)

    async def write(self, key: str, data: bytes) -> None:
        """Store data as a file, creating parent directories as needed.

        On the host filesystem, writes are atomic via write-to-temp-then-rename.

        Args:
            key: Opaque string key identifying the value.
            data: Raw bytes to persist.

        Raises:
            StorageError: If the write fails.
        """
        normalized = _normalize_key(key)
        path = self._path_for(normalized)

        try:
            if self._sandbox is not None:
                await self._sandbox.write_file(path, data)
                return

            parent = os.path.dirname(path)
            os.makedirs(parent, exist_ok=True)

            tmp_path = os.path.join(parent, f"{_TMP_MARKER}_{uuid.uuid4().hex}")
            try:
                with open(tmp_path, "wb") as f:
                    f.write(data)
                os.replace(tmp_path, path)
            except BaseException:
                try:
                    os.unlink(tmp_path)
                except OSError:
                    pass
                raise
        except StorageError:
            raise
        except Exception as error:
            raise StorageError(f"Failed to write '{key}'") from error

    async def read(self, key: str) -> bytes | None:
        """Read the file corresponding to key.

        Args:
            key: The key to read.

        Returns:
            The file contents as bytes, or None if the file does not exist.

        Raises:
            StorageError: If the read fails for a reason other than a missing file.
        """
        normalized = _normalize_key(key)
        path = self._path_for(normalized)

        try:
            if self._sandbox is not None:
                return await self._sandbox.read_file(path)

            with open(path, "rb") as f:
                return f.read()
        except (FileNotFoundError, NotADirectoryError):
            return None
        except StorageError:
            raise
        except Exception as error:
            raise StorageError(f"Failed to read '{key}'") from error

    async def delete(self, key: str) -> None:
        """Delete the file corresponding to key. No-op if it does not exist.

        Args:
            key: The key to delete.

        Raises:
            StorageError: If the delete fails.
        """
        normalized = _normalize_key(key)
        path = self._path_for(normalized)

        try:
            if self._sandbox is not None:
                try:
                    await self._sandbox.remove_file(path)
                except (FileNotFoundError, NotADirectoryError):
                    pass
                return

            try:
                os.unlink(path)
            except (FileNotFoundError, NotADirectoryError):
                pass
        except StorageError:
            raise
        except Exception as error:
            raise StorageError(f"Failed to delete '{key}'") from error

    async def list(self, query: str = "") -> builtins.list[str]:
        """List keys matching the given prefix by walking the directory tree.

        Args:
            query: A prefix string to filter keys. Empty string matches all.

        Returns:
            Matching keys sorted ascending.

        Raises:
            StorageError: If the listing fails.
        """
        prefix = _normalize_prefix(query)

        try:
            if self._sandbox is not None:
                keys = await self._list_keys_sandbox(prefix)
            else:
                keys = self._list_keys_host(prefix)
            return sorted(k for k in keys if k.startswith(prefix))
        except StorageError:
            raise
        except Exception as error:
            raise StorageError(f"Failed to list keys with prefix '{query}'") from error

    def namespace(self, prefix: str) -> _NamespacedStorage:
        """Return a view of this storage with all keys prefixed.

        The returned view preserves ``for_sandbox`` via delegation to the
        underlying storage, so sandbox routing works even when storage is
        pre-namespaced before being passed to a plugin.

        Args:
            prefix: Prefix to prepend to all keys.

        Returns:
            A namespaced storage view.
        """
        return _NamespacedStorage(self, prefix)

    def _path_for(self, key: str) -> str:
        """Map a normalized key to a filesystem path."""
        return os.path.join(self._base_dir, key)

    def _list_keys_host(self, prefix: str) -> builtins.list[str]:
        """Recursively walk the base directory to find all stored keys."""
        base = Path(self._base_dir)

        narrow_dir = base
        if prefix:
            parts = prefix.rstrip("/").split("/")
            for part in parts[:-1]:
                candidate = narrow_dir / part
                if candidate.is_dir():
                    narrow_dir = candidate
                else:
                    break

        keys: builtins.list[str] = []
        if not narrow_dir.exists():
            return keys

        for dirpath, _, filenames in os.walk(narrow_dir):
            for filename in filenames:
                if _TMP_MARKER in filename:
                    continue
                full_path = os.path.join(dirpath, filename)
                rel = os.path.relpath(full_path, self._base_dir)
                key = rel.replace(os.sep, "/")
                keys.append(key)

        return keys

    async def _list_keys_sandbox(self, prefix: str) -> builtins.list[str]:
        """List keys via sandbox file listing."""
        base = Path(self._base_dir)

        keys: builtins.list[str] = []
        await self._walk_sandbox(base, keys)
        return keys

    async def _walk_sandbox(self, directory: Path, keys: builtins.list[str]) -> None:
        """Recursively walk sandbox directories to collect keys."""
        try:
            entries = await self._sandbox.list_files(str(directory))  # type: ignore[union-attr]
        except (FileNotFoundError, NotADirectoryError):
            return

        for entry in entries:
            if _TMP_MARKER in entry.name:
                continue
            full_path = directory / entry.name
            if entry.is_dir:
                await self._walk_sandbox(full_path, keys)
            else:
                rel = os.path.relpath(str(full_path), self._base_dir)
                keys.append(rel.replace(os.sep, "/"))

File: strands-ts/src/storage/local-file-storage.ts
import type { Sandbox } from '../sandbox/base.js'
import type { Storage } from './storage.js'

import { StorageError } from '../errors.js'
import { namespace, normalizeKey, normalizePrefix } from './storage.js'

/**
 * Returns true if the error represents a missing or non-directory path (ENOENT or ENOTDIR).
 *
 * @param error - The caught error to inspect
 * @returns Whether the error is a filesystem not-found error
 */
function isNotFoundError(error: unknown): boolean {
  if (error === null || typeof error !== 'object' || !('code' in error)) return false
  return error.code === 'ENOENT' || error.code === 'ENOTDIR'
}

/**
 * Local-filesystem {@link Storage} backend.
 *
 * Persists each key as a file under a base directory, mapping the key's `/` segments
 * onto directory segments. On the host filesystem, writes are atomic (write to a
 * scratch sibling, then rename) so a crash mid-write never leaves a partially written
 * file. When bound to a {@link Sandbox} via {@link forSandbox}, all I/O is routed
 * through that sandbox instead of the host's `node:fs` (atomicity depends on the
 * sandbox implementation).
 *
 * @example
 * ```typescript
 * import { LocalFileStorage } from '@strands-agents/sdk/storage'
 *
 * const storage = new LocalFileStorage('./.strands/')
 * await storage.write('sessions/abc/snapshot.json', bytes)
 * ```
 */
export class LocalFileStorage implements Storage {
  private readonly _baseDir: string
  private readonly _sandbox: Sandbox | undefined

  /**
   * @param baseDir - Root directory under which keys are stored. Defaults to `./.strands/`.
   * @param sandbox - Optional sandbox to route I/O through. Usually set via {@link forSandbox}.
   */
  constructor(baseDir: string = './.strands/', sandbox?: Sandbox) {
    this._baseDir = baseDir
    this._sandbox = sandbox
  }

  /**
   * Returns a storage instance whose I/O is routed through `sandbox`.
   *
   * Instances already bound to a sandbox return themselves unchanged.
   *
   * @param sandbox - Sandbox to route the returned instance's I/O through
   * @returns A new `LocalFileStorage` with the same base directory, routed through `sandbox`
   */
  forSandbox(sandbox: Sandbox): LocalFileStorage {
    if (this._sandbox) return this
    return new LocalFileStorage(this._baseDir, sandbox)
  }

  /**
   * Stores `data` under `key`, overwriting any existing value.
   *
   * @param key - Opaque string key identifying the value
   * @param data - Raw bytes to persist
   * @throws {@link StorageError} if the key is invalid or the write fails
   */
  async write(key: string, data: Uint8Array): Promise<void> {
    const normalized = normalizeKey(key)
    const path = this._pathFor(normalized)
    if (this._sandbox) {
      try {
        await this._sandbox.writeFile(path, data)
      } catch (error: unknown) {
        throw new StorageError(`Failed to write '${normalized}' to sandbox storage`, { cause: error })
      }
      return
    }
    let tmpPath: string | undefined
    try {
      const { mkdir, writeFile, rename } = await import('node:fs/promises')
      const { dirname } = await import('node:path')
      await mkdir(dirname(path), { recursive: true })
      const { randomUUID } = await import('node:crypto')
      tmpPath = `${path}.__strands_tmp_${randomUUID()}`
      await writeFile(tmpPath, data)
      await rename(tmpPath, path)
    } catch (error: unknown) {
      if (tmpPath) {
        const { rm } = await import('node:fs/promises')
        await rm(tmpPath, { force: true }).catch(() => {})
      }
      throw new StorageError(`Failed to write '${normalized}' to local storage`, { cause: error })
    }
  }

  /**
   * Retrieves the bytes previously stored under `key`.
   *
   * @param key - The key to read
   * @returns The stored bytes, or `null` if no value exists for `key`
   * @throws {@link StorageError} if the key is invalid or the read fails
   */
  async read(key: string): Promise<Uint8Array | null> {
    const normalized = normalizeKey(key)
    const path = this._pathFor(normalized)
    if (this._sandbox) {
      try {
        return await this._sandbox.readFile(path)
      } catch (error: unknown) {
        if (isNotFoundError(error)) return null
        throw new StorageError(`Failed to read '${normalized}' from sandbox storage`, { cause: error })
      }
    }
    try {
      const { readFile } = await import('node:fs/promises')
      const content = await readFile(path)
      return new Uint8Array(content)
    } catch (error: unknown) {
      if (isNotFoundError(error)) return null
      throw new StorageError(`Failed to read '${normalized}' from local storage`, { cause: error })
    }
  }

  /**
   * Deletes the value stored under `key`. A no-op if the key does not exist.
   *
   * @param key - The key to delete
   * @throws {@link StorageError} if the key is invalid or the delete fails
   */
  async delete(key: string): Promise<void> {
    const normalized = normalizeKey(key)
    const path = this._pathFor(normalized)
    if (this._sandbox) {
      try {
        await this._sandbox.removeFile(path)
      } catch (error: unknown) {
        if (!isNotFoundError(error)) {
          throw new StorageError(`Failed to delete '${normalized}' from sandbox storage`, { cause: error })
        }
      }
      return
    }
    try {
      const { rm } = await import('node:fs/promises')
      await rm(path, { force: true })
    } catch (error: unknown) {
      throw new StorageError(`Failed to delete '${normalized}' from local storage`, { cause: error })
    }
  }

  /**
   * Lists the keys whose names begin with `prefix`, sorted lexicographically.
   *
   * @param prefix - Key prefix to match. An empty string matches all keys.
   * @returns The matching keys, sorted ascending
   * @throws {@link StorageError} if the prefix is invalid or the listing fails
   */
  async list(prefix: string): Promise<string[]> {
    const normalized = normalizePrefix(prefix)
    const base = this._baseDir.replace(/\/$/, '')
    // Narrow the walk to the deepest directory the prefix fully specifies
    const lastSlash = normalized.lastIndexOf('/')
    const dirPortion = lastSlash >= 0 ? normalized.slice(0, lastSlash) : ''
    const startDir = dirPortion ? `${base}/${dirPortion}` : base
    const keys = this._sandbox
      ? await this._listKeysSandbox(startDir, dirPortion)
      : await this._listKeysHost(startDir, dirPortion)
    return keys.filter((key) => key.startsWith(normalized)).sort()
  }

  private _pathFor(key: string): string {
    const base = this._baseDir.replace(/\/$/, '')
    return `${base}/${key}`
  }

  private async _listKeysHost(dir: string, keyPrefix: string): Promise<string[]> {
    const { readdir } = await import('node:fs/promises')

    const walk = async (walkDir: string, walkPrefix: string): Promise<string[]> => {
      let entries
      try {
        entries = await readdir(walkDir, { withFileTypes: true })
      } catch (error: unknown) {
        if (isNotFoundError(error)) return []
        throw new StorageError(`Failed to list local storage under '${walkPrefix}'`, { cause: error })
      }
      const found: string[] = []
      for (const entry of entries) {
        if (!entry.isDirectory() && entry.name.includes('.__strands_tmp')) continue
        const childKey = walkPrefix ? `${walkPrefix}/${entry.name}` : entry.name
        if (entry.isDirectory()) {
          found.push(...(await walk(`${walkDir}/${entry.name}`, childKey)))
        } else {
          found.push(childKey)
        }
      }
      return found
    }

    return walk(dir, keyPrefix)
  }

  private async _listKeysSandbox(dir: string, keyPrefix: string): Promise<string[]> {
    const sandbox = this._sandbox!

    const walk = async (walkDir: string, walkPrefix: string): Promise<string[]> => {
      let entries
      try {
        entries = await sandbox.listFiles(walkDir)
      } catch (error: unknown) {
        if (isNotFoundError(error)) return []
        throw new StorageError(`Failed to list sandbox storage under '${walkPrefix}'`, { cause: error })
      }
      const found: string[] = []
      for (const entry of entries) {
        if (!entry.isDir && entry.name.includes('.__strands_tmp')) continue
        const childKey = walkPrefix ? `${walkPrefix}/${entry.name}` : entry.name
        if (entry.isDir) {
          found.push(...(await walk(`${walkDir}/${entry.name}`, childKey)))
        } else {
          found.push(childKey)
        }
      }
      return found
    }

    return walk(dir, keyPrefix)
  }

  /**
   * Returns a prefixed view of this storage without mutating the original.
   * The returned view preserves `forSandbox` for single-level namespacing;
   * nested `.namespace()` calls on the view do not carry sandbox routing.
   */
  namespace(prefix: string): Storage & { forSandbox(sandbox: Sandbox): Storage } {
    const view = namespace(this, prefix)
    return {
      ...view,
      forSandbox: (sandbox: Sandbox): Storage => namespace(this.forSandbox(sandbox), prefix),
    }
  }
}
